File: p17.rb
# 21944
# 21124
# Problem 17: Number letter counts

# 	If the numbers 1 to 5 are written out in words: one, two, three, four, five, then there are 3 + 3 + 5 + 4 + 4 = 19 letters used in total.
# 	If all the numbers from 1 to 1000 (one thousand) inclusive were written out in words, how many letters would be used?
# 	NOTE: Do not count spaces or hyphens. For example, 342 (three hundred and forty-two) contains 23 letters and 115 (one hundred and fifteen) contains 20 letters. 
# 	The use of "and" when writing out numbers is in compliance with British usage.

H = {
	0 => "", 
	1 =>  'one', 2 => 'two', 3 => 'three', 4 => 'four', 5 => 'five', 6 => 'six', 7 => 'seven', 8 => 'eight', 9 => 'nine', 
	11 =>  'eleven', 12 => 'twelve', 13 => 'thirteen', 14 => 'fourteen', 15 => 'fifteen', 16 => 'sixteen', 17 => 'seventeen', 18 => 'eighteen', 19 => 'nineteen', 
	10 =>  'ten', 20 => 'twenty', 30 => 'thirty', 40 => 'forty', 50 => 'fifty', 60 => 'sixty', 70 => 'seventy', 80 => 'eighty', 90 => 'ninety', 100 => 'onehundred', 
	1000 => 'onethousand'
}

def transfer(num)
	(1..num).map do |n|
		if H[n]
			H[n]
		else
			div, mod = n.divmod(100)
			if mod == 0
				H[div] + 'hundred'
			elsif H[mod]
				H[div] + 'hundredand' + H[mod]
			else
				div1, mod1 = mod.to_i.divmod(10)
				s = div == 0 ? "" : H[div] + 'hundredand'
				s + H["#{div1}0".to_i] + H[mod1]
			end
		end
	end.join.length
end

p transfer(1000)

File: P17.java
public class P17 {

	static String unitdo[] = {"", "One", "Two", "Three", "Four", "Five",
		"Six", "Seven", "Eight", "Nine", "Ten", "Eleven", "Twelve",
		"Thirteen", "Fourteen", "Fifteen",  "Sixteen", "Seventeen", 
		"Eighteen", "Nineteen"};

	static String tens[] =  {"","Ten","Twenty", "Thirty", "Forty", "Fifty",
		"Sixty", "Seventy", "Eighty","Ninety"};




	public static int digit_counter(int a)
	{
		int digit_count = 0 ;
		while(a>0)
		{ 
			a=a/10;
			digit_count++;

		}

		return digit_count++;

	}

	public static String wordGen(int a)
	{
		String numWord = "";

		int digits = digit_counter(a);

		while(digits >0)
		{
			switch(digits)
			{
				case 1:
				case 2:

					if(a < 20)
					{ 
						numWord+=unitdo[a];

						digits = 0;
					}
					else
						if(a >=20 )
						{


							numWord+= tens[a/10];
							a=a%10;
							if(a ==0)
								digits = 0;
							else
								digits--;
						}
					break;
				case 3:

					numWord+=unitdo[a/100];
					a=a%100;
					if(a%100 == 0)
					{
						numWord+="Hundred";

						digits = 0;
					}
					else
					{
						numWord+="HundredAnd";
						digits--;
					}
					break;
				case 4:

					numWord+=unitdo[a/1000];
					a=a%1000;
					numWord+="thousand";
					if(a%1000 ==0)
						digits = 0;
					else
						digits--;
					break;



			}

		}

		return numWord;

	}
	public static void main(String[] args) {
		int sum = 0;

		String a = "";
		for(int i =1 ; i<=1000; i++)
		{
			a = wordGen(i);
			System.out.println(a);
			sum= sum+a.length();

		}
		System.out.println(sum);
	}

}

